### DISCUSSION TITLE
Problem with latency playing Wave files

### DISCUSSION THREAD

**Discussion by egscholten** [FactoryXCode/MfPack#38] [Q&A] [answered]
Hi,

I'm trying to play wavefiles using MediaEngineClass. So far so good. On my normal soundcard everything is fine. But if I insert my Asio enabled USB Sound interface, things turn very laggy. I takes about 2 or 3 seconds to play a simple wave. If i turn the latency of the soundcard down to 1 ms everything is fine. But above 10ms is unworkable.

Any pointers to how to fix this?

Thanks,
Gert Scholten

**Comment by FactoryXCode** [comment]
Hi,

Sounds like a buffer or driver issue to me.
Maybe using IMFMediaEngineEx would help. This is the extension of  IMFMediaEngine and has more methods to check on streams, latency etc.

Further I found a good explanation about [ASIO Audio Tutorial – Low Latency Audio On Windows](https://www.google.com/url?sa=t&rct=j&q=&esrc=s&source=web&cd=&cad=rja&uact=8&ved=2ahUKEwja-NGo8qL6AhVZ_7sIHZCiC44QwqsBegQIDxAB&url=https%3A%2F%2Fwww.youtube.com%2Fwatch%3Fv%3D5kNEfjnoyec&usg=AOvVaw1u84LYgkRy7Z15lBEO1XDL)

Finally:

If you want to work with low or no latency on audio the MediaEngine is not designed for that. 
Instead you should use the [Windows Audio Session API (WASAPI)](https://learn.microsoft.com/en-us/windows/win32/coreaudio/wasapi)  like the IAudioClient interface. 
The IAudioClient interface communicates directly with audio hardware and thus no or less latency. 
On the other hand, rendering high quality audio via USB produces latency too, generally it's a good idea to 
communicate with hardware instead of software drivers. 
By default Windows has a latency of approximately 100 to 10 ms on modern machines using software drivers.

Regards, Tony.

**Comment by FactoryXCode** [comment, marked as answer]
Recently a XAudio2 sample has been committed.
XAudio2 is the replacement for DirectSound and performs very well with latency.
You can find the XAudio2 sample [here](https://github.com/FactoryXCode/MfPack/tree/Master/MfPack/Samples/XAudio2Player/Example1).

Regards Tony.
